core/network/impl/stream_engine.hpp

```diff
@@ -44,11 +47,40 @@ namespace kagome::network {
    private:
     struct ProtocolDescr {
       std::shared_ptr<ProtocolBase> protocol;
-      std::shared_ptr<Stream> incoming;
-      std::shared_ptr<Stream> outgoing;
-      std::queue<std::function<void(std::shared_ptr<Stream>)>>
+
+      struct {
+        std::shared_ptr<Stream> stream;
+      } incoming;
+
+      struct {
+        std::shared_ptr<Stream> stream;
+        bool reserved = false;
+      } outgoing;
+
+      std::deque<std::function<void(std::shared_ptr<Stream>)>>
           deffered_messages;
+
+      bool hasActiveOutgoing() const {
+        return outgoing.stream && !outgoing.stream->isClosed();
+      }
+
+      bool reserve() {
+        if (outgoing.reserved || hasActiveOutgoing()) return false;
```

Comment:
Shortened form of the if body is also acceptable, but its use is discouraged too for the sake of cleaness and clarity.